crates/rolldown_plugin_dynamic_import_vars/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+    },
+    AstBuilder, VisitMut,
+  },
+  span::SPAN,
+  syntax::operator::{BinaryOperator, UnaryOperator},
+};
+use rolldown_plugin::{HookTransformAstArgs, HookTransformAstReturn, Plugin, SharedPluginContext};
+use std::{
+  borrow::Cow,
+  path::{Path, PathBuf},
+};
+use sugar_path::SugarPath;
+use to_glob::to_glob_pattern;
+mod clone_expr;
+mod should_ignore;
+mod to_glob;
+
+#[derive(Debug)]
+pub struct DynamicImportVarsPlugin {
+  pub error_when_no_files_found: bool,
+}
+
+#[async_trait::async_trait]
+impl Plugin for DynamicImportVarsPlugin {
+  fn name(&self) -> Cow<'static, str> {
+    Cow::Borrowed("dynamic_import_vars")
+  }
+
+  fn transform_ast(
+    &self,
+    _ctx: &SharedPluginContext,
+    mut args: HookTransformAstArgs,
+  ) -> HookTransformAstReturn {
+    args.ast.program.with_mut(|fields| {
+      let ast_builder: AstBuilder = AstBuilder::new(fields.allocator);
+      let mut visitor = DynamicImportVarsVisit {
+        cwd: args.cwd,
+        ast_builder,
+        error_when_no_files_found: self.error_when_no_files_found,
+        helper_decls: ast_builder.vec(),
+        current: 0,
+      };
+      visitor.visit_program(fields.program);
+      if !visitor.helper_decls.is_empty() {
+        fields.program.body.extend(visitor.helper_decls);
+      }
+    });
+    Ok(args.ast)
+  }
+}
+
+pub struct DynamicImportVarsVisit<'ast, 'a> {
+  cwd: &'a PathBuf,
+  ast_builder: AstBuilder<'ast>,
+  error_when_no_files_found: bool,
+  helper_decls: Vec<'ast, Statement<'ast>>,
+  current: usize,
+}
+
+impl<'ast, 'a> VisitMut<'ast> for DynamicImportVarsVisit<'ast, 'a> {
+  #[allow(clippy::too_many_lines)]
+  fn visit_expression(&mut self, expr: &mut Expression<'ast>) {
+    if let Expression::ImportExpression(import_expr) = expr {
+      let pattern = to_glob_pattern(&import_expr.source).unwrap();
+      if let Some(pattern) = pattern {
+        let path = Path::new(self.cwd).join(Path::new(&pattern));
+        let mut files = vec![];
+        if path.is_absolute() {
+          if let Some(path) = path.to_str() {
+            for file in glob(path).unwrap() {
+              let file = file.unwrap().as_path().relative(self.cwd).to_slash_lossy().to_string();
+              files.push(format!("./{file}"));
+            }
+          }
+        }
+
+        assert!(!(self.error_when_no_files_found && files.is_empty()), "No files found in {pattern:?} when trying to dynamically load concatenated string from {:?}", self.cwd);
+
+        let name = format!("__variableDynamicImportRuntime{}__", self.current);
+        let import_arg = import_expr.arguments.first();
+
+        let helper_decl = self.helper_func(&name, files, import_arg);
+        self.helper_decls.push(self.ast_builder.statement_declaration(helper_decl));
+
+        *expr = self.ast_builder.expression_call(
+          import_expr.span,
+          self.ast_builder.vec1(
+            self.ast_builder.argument_expression(clone_expr(self.ast_builder, &import_expr.source)),
```

Comment:
Here the `&import_expr.source` is not be used again, you could take it out form `import_expr` to avoid clone.

Reply:
The `import_expr` is a reference, thus it can't be moved into `Argument` 🤔

Reply:
Here it is a `&mut import_expr`, i will using to move it out.

```
std::mem::replace(
            &mut import_expr.source,
            self.ast_builder.expression_null_literal(SPAN),
          )
```